src/compiler/program.ts

```diff
@@ -2532,8 +2532,8 @@ namespace ts {
             }
 
             if (options.declarationDir) {
-                if (!options.declaration) {
-                    createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "declarationDir", "declaration");
+                if (!(options.declaration || options.composite)) {
+                    createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "declarationDir", options.declaration ? "declaration" : "composite");

```

Comment:
why use a different diagnostic? the new one seems wrong here

Reply:
I guess you don't need to change the error message at all since `declaration` is implicitly enabled by `composite`